File: adventofcode/day3/bin/day3_part2.php
<?php 

class GridRow { 
  var $values;
  function __construct( ) {
    $this->values = array();
  }
}
class Grid { 
  var $cols;
  var $minx;
  var $maxx;
  var $miny;
  var $maxy;
  var $turtle_limit;
  var $first_largest;
  function __construct( ) {
    $this->cols = array();
    $this->minx = 0;
    $this->maxx = 0;
    $this->miny = 0;
    $this->maxy = 0;
    $this->turtle_limit = 0;
    $this->first_largest = 0;
  }
  function getValue( $x , $y ) {
    if ( array_key_exists($y , $this->cols ) ) {
      $col = $this->cols[$y];
      if ( array_key_exists($x , $col->values ) ) {
        return ($col->values[$x]);
      }
    }
    return 0;
  }
  function setValue( $x , $y , $value ) {
    if ( ($value > $this->turtle_limit) && ($this->first_largest == 0) ) {
      $this->first_largest = $value;
    }
    if ( array_key_exists($y , $this->cols ) ) {
      $col = $this->cols[$y];
      $col->values[$x] = $value;
    } else {
      $newCol =  new GridRow();
      $this->cols[$y] = $newCol;
      $newCol->values[$x] = $value;
    }
    if ( $x < $this->minx ) {
      $this->minx = $x;
    }
    if ( $x > $this->maxx ) {
      $this->maxx = $x;
    }
    if ( $y < $this->miny ) {
      $this->miny = $y;
    }
    if ( $y > $this->maxy ) {
      $this->maxy = $y;
    }
  }
  function getAdjacentSum( $x , $y ) {
    $v = (function ($dx, $dy) use ( &$x,  &$y) {
      return ($this)->getValue(($x + $dx), ($y + $dy));
      // captured var x
      // captured var y
    });
    return ((((((call_user_func($v, -1, -1) + call_user_func($v, -1, 0)) + call_user_func($v, -1, 1)) + call_user_func($v, 0, -1)) + call_user_func($v, 0, 1)) + call_user_func($v, 1, -1)) + call_user_func($v, 1, 0)) + call_user_func($v, 1, 1);
  }
  function printGrid() {
    $xx = $this->minx;
    $yy = $this->miny;
    while ($yy <= $this->maxy) {
      $xx = $this->minx;
      $row = array();
      while ($xx <= $this->maxx) {
        array_push($row, ($this)->getValue($xx, $yy));
        $xx = $xx + 1;
      }
      echo( implode(" ", operatorsOf::map_2($row, (function ($item, $index)  {
        return strval($item);
      }))) . "\n");
      $yy = $yy + 1;
    }
  }
}
class day_three_part_two { 
  function __construct( ) {
  }
  public static function distance( $data ) {
    $myGrid =  new Grid();
    $myGrid->turtle_limit = $data;
    $myGrid->setValue(0, 0, 1);
    $i = 0;
    $j = 0;
    $moveTurtle = (function ($dx, $dy, $steps) use ( &$i,  &$j,  &$myGrid) {
      $cnt = $steps;
      while ($cnt > 0) {
        $i = $i + $dx;
        $j = $j + $dy;
        $sum = $myGrid->getAdjacentSum($i, $j);
        $myGrid->setValue($i, $j, $sum);
        $cnt = $cnt - 1;
      }
      // captured var i
      // captured var j
      // captured var myGrid
    });
    $step = 2;
    while ($myGrid->first_largest == 0) {
      call_user_func($moveTurtle, 1, 0, 1);
      call_user_func($moveTurtle, 0, -1, $step - 1);
      call_user_func($moveTurtle, -1, 0, $step);
      call_user_func($moveTurtle, 0, 1, $step);
      call_user_func($moveTurtle, 1, 0, $step);
      $step = $step + 2;
    }
    $myGrid->printGrid();
    echo( "the first largest was " . $myGrid->first_largest . "\n");
    return $myGrid->turtle_limit;
  }
}
/* static PHP main routine */
day_three_part_two::distance(289326);
class operatorsOf { 
  function __construct( ) {
  }
  public static function map_2( $__self , $cb ) {
    /** unused:  $__len = count($__self)   **/ ;
    $res = array();
    for ( $i = 0; $i < count($__self); $i++) {
      $it = $__self[$i];
      array_push($res, call_user_func($cb, $it, $i));
    }
    return $res;
  }
}

File: legacy/adventofcode/day3/bin/day3_part2.swift
func ==(l: GridRow, r: GridRow) -> Bool {
  return l === r
}
class GridRow : Equatable  { 
  var values : [Int:Int] = [Int:Int]()
}
func ==(l: Grid, r: Grid) -> Bool {
  return l === r
}
class Grid : Equatable  { 
  var cols : [Int:GridRow] = [Int:GridRow]()
  var minx : Int = 0
  var maxx : Int = 0
  var miny : Int = 0
  var maxy : Int = 0
  var turtle_limit : Int = 0
  var first_largest : Int = 0
  func getValue(x : Int, y : Int) -> Int {
    if ( self.cols[y] != nil ) {
      let col : GridRow? = self.cols[y]
      if ( col!.values[x] != nil ) {
        return (col!.values[x])!;
      }
    }
    return 0;
  }
  func setValue(x : Int, y : Int, value : Int) -> Void {
    if ( (value > self.turtle_limit) && (self.first_largest == 0) ) {
      self.first_largest = value;
    }
    if ( self.cols[y] != nil ) {
      let col : GridRow? = self.cols[y]
      col!.values[x] = value;
    } else {
      let newCol : GridRow = GridRow()
      self.cols[y] = newCol;
      newCol.values[x] = value;
    }
    if ( x < self.minx ) {
      self.minx = x;
    }
    if ( x > self.maxx ) {
      self.maxx = x;
    }
    if ( y < self.miny ) {
      self.miny = y;
    }
    if ( y > self.maxy ) {
      self.maxy = y;
    }
  }
  func getAdjacentSum(x : Int, y : Int) -> Int {
    let v : (( _ : Int,  _ : Int) -> Int) = ({ (dx, dy) ->  Int in 
      return (self).getValue(x : (x + dx), y : (y + dy));
      // captured var x
      // captured var y
    })
    return ((((((v(-1, -1) + v(-1, 0)) + v(-1, 1)) + v(0, -1)) + v(0, 1)) + v(1, -1)) + v(1, 0)) + v(1, 1);
  }
  func printGrid() -> Void {
    var xx : Int = self.minx
    var yy : Int = self.miny
    while (yy <= self.maxy) {
      xx = self.minx;
      var row : [Int] = [Int]()
      while (xx <= self.maxx) {
        row.append((self).getValue(x : xx, y : yy))
        xx = xx + 1;
      }
      print(operatorsOf.map_2(__self : row, cb : ({ (item, index) ->  String in 
        return String(item);
      })).joined(separator:" "))
      yy = yy + 1;
    }
  }
}
func ==(l: day_three_part_two, r: day_three_part_two) -> Bool {
  return l === r
}
class day_three_part_two : Equatable  { 
  class func distance(data : Int) -> Int {
    let myGrid : Grid = Grid()
    myGrid.turtle_limit = data;
    myGrid.setValue(x : 0, y : 0, value : 1)
    var i : Int = 0
    var j : Int = 0
    let moveTurtle : (( _ : Int,  _ : Int,  _ : Int) -> Void) = ({ (dx, dy, steps) ->  Void in 
      var cnt : Int = steps
      while (cnt > 0) {
        i = i + dx;
        j = j + dy;
        let sum : Int = myGrid.getAdjacentSum(x : i, y : j)
        myGrid.setValue(x : i, y : j, value : sum)
        cnt = cnt - 1;
      }
      // captured var i
      // captured var j
      // captured var myGrid
    })
    var step : Int = 2
    while (myGrid.first_largest == 0) {
      moveTurtle(1, 0, 1);
      moveTurtle(0, -1, step - 1);
      moveTurtle(-1, 0, step);
      moveTurtle(0, 1, step);
      moveTurtle(1, 0, step);
      step = step + 2;
    }
    myGrid.printGrid()
    print("the first largest was " + String(myGrid.first_largest))
    return myGrid.turtle_limit;
  }
}
func ==(l: operatorsOf, r: operatorsOf) -> Bool {
  return l === r
}
class operatorsOf : Equatable  { 
  class func map_2(__self : [Int], cb :   @escaping  (( _ : Int,  _ : Int) -> String)) -> [String] {
    /** unused:  let __len : Int = __self.count   **/ 
    var res : [String] = [String]()
    for ( i , it ) in __self.enumerated() {
      res.append(cb(it, i))
    }
    return res;
  }
}
func __main__swift() {
  _ = day_three_part_two.distance(data : 289326)
}
// call the main function
__main__swift()
